### PR TITLE
*: Cut 2.29.0-rc.0

### PR DESCRIPTION
**Pull request by brancz**
Signed-off-by: Frederic Branczyk <fbranczyk@gmail.com>

@juliusv @roidelapluie @simonpasquier @bwplotka @codesome @SuperQ

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,31 @@
...
+* [FEATURE] promtool: Add feature flags flag `--enable-feature`. #8958
+* [FEATURE] promtool: Add service discovery configuration validation. #8950
+* [ENHANCEMENT] Allow filtering with matchers in label-names API. #9083
+* [ENHANCEMENT] Add checkpoint replay to write-ahead-log replay status. #8898
+* [ENHANCEMENT] Log when total symbol size exceeds 2^32 bytes, causing compaction to fail. #9104
+* [ENHANCEMENT] Reduce blocking of outgoing remote write requests from series garbage collection. #9109
+* [ENHANCEMENT] Improve write-ahead-log decoding performance. #9106
+* [ENHANCEMENT] Improve write performance in TSDB. #9061
+* [ENHANCEMENT] Allow configuring `max_samples_per_send` for remote write metadata. #8959
+* [ENHANCEMENT] Allow charset in content-type in HTTP discovery. #9061
```
Comment (roidelapluie) [inline CHANGELOG.md:19]: This was already there in 2.28.1

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,31 @@
+## 2.29.0-rc.0 / 2021-07-29
+
+* [CHANGE] Remove experimental tag on `--storage.tsdb.allow-overlapping-blocks` flag. #9117
+* [CHANGE] Remove experimental tag on `--storage.tsdb.retention.size` flag. #8728
+* [FEATURE] Add Kuma service discovery. #8844
+* [FEATURE] Add `__meta_gce_interface_ipv4_<name>` meta label to GCE discovery. #8978
+* [FEATURE] Add `__meta_ec2_availability_zone_id` meta label to EC2 discovery. #8978
+* [FEATURE] Add `__meta_hetzner_hcloud_labelpresent_<labelname>` meta label to Hetzner discovery. #9028
```
Comment (roidelapluie) [inline CHANGELOG.md:8]:
```suggestion
* [ENHANCEMENT] Add `__meta_gce_interface_ipv4_<name>` meta label to GCE discovery. #8978
* [ENHANCEMENT] Add `__meta_ec2_availability_zone_id` meta label to EC2 discovery. #8978
* [ENHANCEMENT] Add `__meta_hetzner_hcloud_labelpresent_<labelname>` meta label to Hetzner discovery. #9028
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,3 +1,31 @@
...
+* [FEATURE] Allow configuring exemplar storage via file and make it reloadable. #8974
+* [FEATURE] promtool: Add feature flags flag `--enable-feature`. #8958
+* [FEATURE] promtool: Add service discovery configuration validation. #8950
```
Comment (roidelapluie) [inline CHANGELOG.md:11]:
```suggestion
* [FEATURE] promtool: Add file_sd files configuration validation. #8950
```

Comment (roidelapluie) [reply CHANGELOG.md:11]: We should call out that this is file_sd.

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,31 @@
...
+* [ENHANCEMENT] Allow filtering with matchers in label-names API. #9083
```
Comment (roidelapluie) [inline CHANGELOG.md:12]: This is not user visible, I would not have it here.

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,31 @@
...
+* [ENHANCEMENT] Add checkpoint replay to write-ahead-log replay status. #8898
```
Comment (roidelapluie) [inline CHANGELOG.md:13]: Should this be mentioned?

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,31 @@
...
+* [ENHANCEMENT] Log when total symbol size exceeds 2^32 bytes, causing compaction to fail. #9104
```
Comment (roidelapluie) [inline CHANGELOG.md:14]: Bugfix? Now we do not do the compation anymore.

File: CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,31 @@
...
+* [ENHANCEMENT] Reduce blocking of outgoing remote write requests from series garbage collection. #9109
+* [ENHANCEMENT] Improve write-ahead-log decoding performance. #9106
+* [ENHANCEMENT] Improve write performance in TSDB. #9061
```
Comment (roidelapluie) [inline CHANGELOG.md:17]:
```suggestion
* [ENHANCEMENT] Improve append performance in TSDB by reducing mutexes usage. #9061
```

TSDB is a lot, I would be more specific.

File: CHANGELOG.md (hunk 2 above)
Comment (roidelapluie) [reply CHANGELOG.md:8]: I would count these as enhancement.

**Review by roidelapluie — COMMENTED** [review]
My English is failing me today, so do not take my recommendations as verbatim, but more like comments.

Thanks!

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,3 +1,31 @@
+## 2.29.0-rc.0 / 2021-07-29
+
+* [CHANGE] Remove experimental tag on `--storage.tsdb.allow-overlapping-blocks` flag. #9117
+* [CHANGE] Remove experimental tag on `--storage.tsdb.retention.size` flag. #8728
+* [FEATURE] Add Kuma service discovery. #8844
+* [FEATURE] Add `__meta_gce_interface_ipv4_<name>` meta label to GCE discovery. #8978
```
Comment (juliusv) [inline CHANGELOG.md:6]: For new labels in SDs and other minor new features, I also had them as `FEATURE` initially for the last release, but @roidelapluie's opinion was that they should be treated as an `ENHANCEMENT` instead, see discussion in https://github.com/prometheus/prometheus/pull/8954#issuecomment-863285982.

At least for consistency, we would want to downgrade them to `ENHANCEMENT` here too, but I'm also not sure if we should keep that policy.

File: CHANGELOG.md (hunk 4 above)
Comment (juliusv) [inline CHANGELOG.md:12]: This on the other hand is something I'd be tempted to call a `FEATURE` due to its potentially big significance to API users, but @roidelapluie would maybe disagree with me again :)

Comment (roidelapluie) [reply CHANGELOG.md:12]: This is not a change for users, just now we filter at TSDB layer instead of API layer, if I remember correctly.

**Comment by roidelapluie** [comment]
You should you edit this PR and point it to a release-2.29 branch.

File: CHANGELOG.md [hunk 9]
```diff
@@ -1,3 +1,31 @@
...
+* [ENHANCEMENT] Improve write performance in TSDB. #9061
+* [ENHANCEMENT] Allow configuring `max_samples_per_send` for remote write metadata. #8959
+* [ENHANCEMENT] Allow charset in content-type in HTTP discovery. #9061
+* [ENHANCEMENT] promtool: Add compaction efficiency to `promtool tsdb analyze` reports. #8940
+* [ENHANCEMENT] promtool: Allow configuring max block duration for backfilling via `--max-block-duration` flag. #8919
+* [ENHANCEMENT] UI: Add sorting and filtering to flags page. #8988
+* [ENHANCEMENT] UI: Improve alerts page rendering performance. #9005
+* [BUGFIX] Fix incorrect `target_limit` reloading of zero value. #9120
+* [BUGFIX] Fix head GC and pending readers race condition. #9081
+* [BUGFIX] Fix timestamp handing in OpenMetrics parser. #9008
```
Comment (juliusv) [inline CHANGELOG.md:26]:
```suggestion
* [BUGFIX] Fix timestamp handling in OpenMetrics parser. #9008
```

Comment (roidelapluie) [reply CHANGELOG.md:26]: Fix invalid timestamp handling in OpenMetrics parser. #9008

File: CHANGELOG.md (hunk 4 above)
Comment (juliusv) [reply CHANGELOG.md:12]: I see, I thought it was an actual new feature, not just an improvement of the implementation. Although the pull request (https://github.com/prometheus/prometheus/pull/9083) mentions big resource savings (which we usually do surface in changelogs), but probably the label names endpoint is not significant enough for this.

**Comment by roidelapluie** [comment]
if this is merged to main, we need to include #9097

**Comment by brancz** [comment]
Good catch, I missed #9097

File: CHANGELOG.md (hunk 4 above)
Comment (brancz) [reply CHANGELOG.md:12]: https://github.com/prometheus/prometheus/pull/9133#discussion_r678993835 I think I'd want to include it for API users.

File: CHANGELOG.md (hunk 1 above)
Comment (brancz) [reply CHANGELOG.md:19]: ah thanks for catching!

File: CHANGELOG.md (hunk 5 above)
Comment (brancz) [reply CHANGELOG.md:13]: Yeah, I wasn't too sure either. I'll skip it.

File: CHANGELOG.md (hunk 7 above)
Comment (brancz) [reply CHANGELOG.md:17]: fair enough

File: CHANGELOG.md (hunk 6 above)
Comment (brancz) [reply CHANGELOG.md:14]: makes sense, changing

**Comment by roidelapluie** [comment]
> Good catch, I missed #9097

it was merged after this pull request

**Comment by brancz** [comment]
Ah makes sense. What are your thoughts on https://github.com/prometheus/prometheus/pull/8970? If it looks good to you I'm happy including it in this release.

**Comment by LeviHarrison** [comment]
Is https://github.com/prometheus/prometheus/pull/8977 in there?

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,31 @@
+## 2.29.0-rc.0 / 2021-07-29
+
+* [CHANGE] Remove experimental tag on `--storage.tsdb.allow-overlapping-blocks` flag. #9117
+* [CHANGE] Remove experimental tag on `--storage.tsdb.retention.size` flag. #8728
```
Comment (codesome) [inline CHANGELOG.md:4]: I guess we usually point to the PRs that made the change?
```suggestion
* [CHANGE] Remove experimental tag on `--storage.tsdb.retention.size` flag. #9004
```

**Comment by brancz** [comment]
Looks like I missed that one @LeviHarrison, adding!

File: CHANGELOG.md (hunk 4 above)
Comment (codesome) [reply CHANGELOG.md:12]: And to be clear, this only changes the Go interface with some internal enhancements, while the http API behaviour is same as before (we support the matchers in http API for label names already)

Comment (brancz) [reply CHANGELOG.md:12]: Ooooooh, I didn't realize this wasn't added to the HTTP API. Thanks for clarifying, yes then it shouldn't be included.

File: CHANGELOG.md (hunk 10 above)
Comment (bwplotka) [reply CHANGELOG.md:4]: +1

**Review by bwplotka — COMMENTED** [review]
Look good to me module all comments made already , thanks 👍🏽

**Comment by LeviHarrison** [comment]
https://github.com/prometheus/exporter-toolkit/pull/58 and https://github.com/prometheus/prometheus/pull/9123 fixed a bug where we weren't accepting a valid client auth type (https://github.com/prometheus/prometheus/issues/9066). It is user-visible, so maybe it should be in the changelog.

**Comment by brancz** [comment]
Good catch @LeviHarrison, added!

File: CHANGELOG.md [hunk 11]
```diff
@@ -1,3 +1,31 @@
+## 2.29.0-rc.0 / 2021-07-29
+
+* [CHANGE] Remove experimental tag on `--storage.tsdb.allow-overlapping-blocks` flag. #9117
+* [CHANGE] Remove experimental tag on `--storage.tsdb.retention.size` flag. #9004
```
Comment (SuperQ) [inline CHANGELOG.md:4]: I wonder if this could be written as "Promote flag `--...` to stable"

Comment (brancz) [reply CHANGELOG.md:4]: agreed

**Review by roidelapluie — APPROVED** [review]

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,3 +1,32 @@
+## 2.29.0-rc.0 / 2021-07-29
+
+* [CHANGE] Promote `--storage.tsdb.allow-overlapping-blocks` flag to stable. #9117
+* [CHANGE] Promote `--storage.tsdb.retention.size` flag to stable. #9004
+* [FEATURE] Add Kuma service discovery. #8844
+* [FEATURE] Add `present_over_time` PromQL function. #9097
+* [FEATURE] Allow configuring exemplar storage via file and make it reloadable. #8974
```
Comment (bboreham) [inline CHANGELOG.md:7]: I clicked over to see how the "storage via file" worked, only to realise it's "configuring via file".
So you might want to reword for clarity.

**Pull request merged**